Multi-Host MeshDevice Support

cfjchu opened the pull request:
### Ticket
[Link to Github Issue
](https://github.com/tenstorrent/tt-metal/issues/20899)

### Problem description
`MeshDevice` requires an upgrade to support a distributed virtualization across multiple hosts.

### What's changed
This PR enables users to create a `MeshDevice` over a mesh of devices that may span multiple hosts. This is the initial infrastructure used to give users this ability.

<img width="814" height="717" alt="image" src="https://github.com/user-attachments/assets/de83c3d2-368a-48c2-93ef-1a0a4c669ee3" />

## Implementation View

Much of the infrastructure is fairly non-invasive from the API point of view. Users work with a global shapes as though they have the full 8x8 mesh available. Underneath, the dispatch of data/work is done in a distributed way with each host responsible for its own local devices.

<img width="751" height="686" alt="image" src="https://github.com/user-attachments/assets/403b064d-d42e-45b5-8ebe-730ed540d5a7" />

Note: Changes here are only for `MeshDevice` -- there are follow-up changes for `MeshCommandQueue`, `MeshBuffer`, etc in a subsequent PR.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions/runs/16215125337
- [x] T3K Unit Tests: https://github.com/tenstorrent/tt-metal/actions/runs/16215122960

Re-launching new tests after addressing PR feedback:
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions/runs/16231748153
- [x] T3K Unit Tests: https://github.com/tenstorrent/tt-metal/actions/runs/16231744346
- [x] T3K Frequent: https://github.com/tenstorrent/tt-metal/actions/runs/16231846469
- [x] T3K Demo: https://github.com/tenstorrent/tt-metal/actions/runs/16231844258

cfjchu committed:
Preliminary support for multi-host distributed mesh

This allows multiple hosts to collaborate on a single logical mesh
while each host operates only on its physically connected devices.

In the context of a multi-host MeshDevice, each host is responsible for
device management of local devices of the larger distributed mesh.

Key changes:
- Add DistributedMeshConfig to track global shape and local host position
  with validation to ensure local mesh fits within global bounds
- Create CoordinateTranslator utility to centralize global<->local
  coordinate translation logic with comprehensive validation
- Tweak SystemMesh to build translation map only for locally accessible devices
- MeshDevice translates global coordinates to local before device access
- Add `ControlPlane::get_local_mesh_offset` query to retrieve offset
  into locally bound MeshId, HostRankId

omilyutin-tt commented on tt_metal/api/tt-metalium/system_mesh.hpp:
```diff
@@ -32,9 +32,12 @@ class SystemMesh {
     SystemMesh& operator=(SystemMesh&&) = delete;
 
-    // Returns the shape of the system mesh
+    // Returns the shape of the system mesh; this is the global mesh shape in distributed context
```
nit: drop `get_` here for consistency with `local_shape`

cfjchu replied:
done

omilyutin-tt commented on tt_metal/distributed/mesh_device.cpp:
```diff
@@ -244,8 +259,10 @@ std::map<int, std::shared_ptr<MeshDevice>> MeshDevice::create_unit_meshes(
         device_ids, l1_small_size, trace_region_size, num_command_queues, worker_l1_size, dispatch_core_config);
     MeshContainer<IDevice*> devices(MeshShape(1, device_ids.size()), scoped_devices->root_devices());
+
+    auto distributed_mesh_config = DistributedMeshConfigFactory::create_from_control_plane(MeshDeviceConfig(MeshShape(1, device_ids.size())));
```
Will this work here? E.g. if we have 2x galaxy 8x8, trying to reshape to 1x64 will yield you an interesting configuration where first 8 devices are owned by one host, next 8 devices are owned by another host, etc. 

We will need to extend the abstraction of local/global, but for now, wdyt about disallowing reshapes of multi-host meshes?

@TT-BrianLiu

omilyutin-tt replied:
Also it might be problematic to add global / local shapes on top of mesh device view. Is it possible to extend the view class instead?

cfjchu replied:
explicitly prohibiting reshape operation for multi-host mesh as we talked about.

omilyutin-tt commented on tt_metal/distributed/coordinate_translation.hpp:
```diff
@@ -26,7 +26,8 @@ class PhysicalMeshCoordinate {
     using MeshId = tt::tt_fabric::MeshId;
     PhysicalMeshCoordinate() = delete;
-    PhysicalMeshCoordinate(chip_id_t mesh_id, chip_id_t chip_id) : mesh_id_(MeshId(mesh_id)), chip_id_(chip_id) {}
+    PhysicalMeshCoordinate(chip_id_t mesh_id, chip_id_t chip_id)
```
Possible to accept MeshId in ctor directly?

cfjchu committed:
SystemMesh::get_shape() -> SystemMesh::shape()

cfjchu committed:
shape

cfjchu committed:
address comments

cfjchu committed:
Address comments + apply suggestions

cfjchu commented on tt_metal/distributed/coordinate_translator.hpp:
```diff
@@ -0,0 +1,40 @@
+
+// Utility class for translating between global and local coordinate systems
+// in a distributed mesh environment where each host manages a portion of the global mesh.
+class CoordinateTranslator {
```
Agree, need better name for `CoordinateTranslator`. I wanted this to support going between both global and local. Any other suggestions?

cfjchu replied:
I landed on `DistributedCoordinateSystem`

cfjchu commented on tt_metal/distributed/coordinate_translator.hpp:
```diff
@@ -0,0 +1,40 @@
+    // Check if a global coordinate is accessible locally
+    bool is_local_coordinate(const MeshCoordinate& global_coord) const;
+
+    // Translate global coordinate to local, or fatal error if out of bounds
+    MeshCoordinate translate_or_fatal(const MeshCoordinate& global_coord) const;
```
I'm fine with that. Done.

cfjchu committed:
simplication wip

cfjchu committed:
add unit tests

omilyutin-tt commented on tt_metal/distributed/system_mesh.cpp:
```diff
@@ -19,61 +21,98 @@
+}
+
+bool SystemMesh::Impl::is_local_coordinate(const MeshCoordinate& coord) const {
+    if (!physical_coordinates_.coord_range().contains(coord)) {
```
Would this be ok to TT_FATAL on?

cfjchu replied:
this is removed now.

omilyutin-tt commented on tt_metal/api/tt-metalium/maybe_remote.hpp:
```diff
@@ -0,0 +1,208 @@
+    
+    // Simple pattern matching for local/remote cases
+    template <typename LocalFunc, typename RemoteFunc>
+    [[nodiscard]] auto when(LocalFunc&& on_local, RemoteFunc&& on_remote) const -> decltype(auto);
```
Nice :)

omilyutin-tt commented on tt_metal/distributed/mesh_device_view.cpp:
```diff
@@ -223,6 +266,57 @@ std::vector<IDevice*> MeshDeviceView::get_ring_devices() const {
+    return extract_locals(devices_.values());
+}
+
+MeshCoordinate MeshDeviceView::local_offset() const {
```
This method and the one below - will it actually work in face of reshapes? E.g. 8x8 dual galaxy reshaped to 1x64 would have strided devices, both methods would return something odd

cfjchu committed:
Refactor distributed coordinate system and add MaybeRemote tests

- Replace CoordinateTranslator with DistributedCoordinateSystem for clearer separation of concerns
- Remove DistributedMeshConfigFactory in favor of direct configuration
- Add comprehensive unit tests for MaybeRemote functionality
- Update MeshDeviceView and related classes to use new coordinate system

cfjchu committed:
prohibit reshape for multi-host mesh

cfjchu commented on tt_metal/fabric/control_plane.cpp:
```diff
@@ -371,7 +371,7 @@ ControlPlane::ControlPlane(
     const std::map<FabricNodeId, chip_id_t>& logical_mesh_chip_id_to_physical_chip_id_mapping) {
     this->logical_mesh_chip_id_to_physical_chip_id_mapping_ = logical_mesh_chip_id_to_physical_chip_id_mapping;
-    this->validate_mesh_connections();
+    //this->validate_mesh_connections();
```
Yes, this was just to unblock some multi-host testing on my end while Allan completes his changes. I'll revert this.

cfjchu committed:
more cleanup

cfjchu committed:
cleanup

cfjchu committed:
use MaybeRemote::when()

cfjchu committed:
clang-format

cfjchu committed:
Update SPDX licence for relevant files

cfjchu committed:
fix CI failures

cfjchu committed:
fix t3k regression tests

ayerofieiev-tt commented on tt_metal/api/tt-metalium/mesh_coord.hpp:
```diff
@@ -372,6 +373,68 @@ class MeshContainer {
+ * (on other hosts) and some are local. The values are wrapped in MaybeRemote<T> to allow for easy distinction between
+ * local and remote values.
+ *
+ * @tparam T The type of values stored (will be wrapped in MaybeRemote<T>)
```
not sure if tparam is intentional for `template param`

cfjchu replied:
yes intentional

ayerofieiev-tt commented on tt_metal/distributed/distributed_coordinate_system.cpp:
```diff
@@ -0,0 +1,93 @@
+DistributedCoordinateSystem DistributedCoordinateSystem::from_control_plane() {
+    auto& control_plane = tt::tt_metal::MetalContext::instance().get_control_plane();
+
+    auto global_shape = control_plane.get_physical_mesh_shape(
+        control_plane.get_local_mesh_id_bindings()[0], tt::tt_fabric::MeshScope::GLOBAL);
+
+    auto local_shape = control_plane.get_physical_mesh_shape(
+        control_plane.get_local_mesh_id_bindings()[0],
+        tt::tt_fabric::MeshScope::LOCAL);
+    auto local_offset = control_plane.get_local_mesh_offset();
```
(nit)
```suggestion
    auto local_offset = control_plane.get_local_mesh_offset();

        const auto& bindings = control_plane.get_local_mesh_id_bindings()[0]
    auto global_shape = control_plane.get_physical_mesh_shape(bindings, tt::tt_fabric::MeshScope::GLOBAL);
    auto local_shape = control_plane.get_physical_mesh_shape(
        bindings, tt::tt_fabric::MeshScope::LOCAL);

```

cfjchu replied:
done

ayerofieiev-tt commented on tt_metal/distributed/distributed_coordinate_system.hpp:
```diff
@@ -0,0 +1,50 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
```
are headers in this folder - public headers?

cfjchu replied:
strictly just `tt_metal/api`

ayerofieiev-tt commented on tt_metal/distributed/mesh_device_view.cpp:
```diff
@@ -33,12 +35,38 @@ std::vector<IDevice*> get_devices_from_coordinates(
-        device_coordinates_.emplace(device->id(), coord);
+
+    // Convert IDevice* to MaybeRemote<IDevice*> and populate
+    for (const auto& [coord, device] : devices) {
```
if you got an IDevice* and its remote - what does it mean? That object is not touchable, it is not legal to query info from it?

cfjchu replied:
Yes, because one host does not have access to the remote device (no direct physical connection) -- each host just manages its own local devices

cfjchu replied:
credit goes to @omilyutin-tt for the idea:
```
template <typename T>
class MaybeRemote {
public:
    using value_type = T;

private:
    std::variant<RemoteDevice, T> value_;
```

cfjchu replied:
and `RemoteDevice` is just an empty struct

jvegaTT reviewed (approved):
Approving on the file I own

dmakoviichuk-tt commented on tt_metal/distributed/system_mesh.cpp:
```diff
@@ -163,16 +200,26 @@ std::vector<chip_id_t> SystemMesh::Impl::get_mapped_physical_device_ids(
-                auto physical_device_id = get_physical_device_id(system_coord);
-                physical_device_ids.push_back(physical_device_id);
-                log_debug(LogMetal, "Logical coordinate: {}, Physical device ID: {}", system_coord, physical_device_id);
+                if (physical_coordinates_.is_local_at(system_coord)) {
```
this and everything at line 215+ looks the same. Can we avoid copypaste?

cfjchu replied:
done.

ayerofieiev-tt commented on tt_metal/distributed/mesh_device_view.cpp:
```diff
@@ -33,12 +35,38 @@ std::vector<IDevice*> get_devices_from_coordinates(
+    for (const auto& [coord, maybe_device] : devices) {
+        devices_.at(coord) = maybe_device;
+
+        maybe_device.when(
+            [this, &coord](const auto& device) { device_coordinates_.emplace(device->id(), coord); },
+            []() { /* skip remote devices */ }
+        );
```
(nit)
`maybe_device.if_local(lambda)`

cfjchu replied:
good idea

cfjchu replied:
Implemented.

tt-rkim reviewed (approved):
Please run T3K unit tests at least
and any pipelines containing model tests that may use meshes (not sure if there are any)

cfjchu commented on tt_metal/distributed/mesh_device_view.cpp:
```diff
@@ -223,6 +271,57 @@ std::vector<IDevice*> MeshDeviceView::get_ring_devices() const {
+        }
+    }
+
+    // If no local devices, return (0, 0) - this shouldn't happen in practice
```
yes, that's more appropriate

cfjchu commented on tt_metal/distributed/mesh_device_view.cpp:
```diff
@@ -223,6 +271,57 @@ std::vector<IDevice*> MeshDeviceView::get_ring_devices() const {
+    }
+
+    if (!min_coord.has_value() || !max_coord.has_value()) {
+        // No local devices - return empty shape
```
yes

dmakoviichuk-tt commented on tt_metal/api/tt-metalium/maybe_remote.hpp:
```diff
@@ -0,0 +1,237 @@
+ * @return std::vector<T> containing only the local values
+ */
+template <typename Container>
+[[nodiscard]] auto extract_locals(const Container& container) {
```
Can we addd concept here:
```
template <typename T>
concept MaybeRemoteLike = requires(T t) {
    typename T::value_type;
    { t.is_local() } -> std::convertible_to<bool>;
    { t.value() } -> std::same_as<typename T::value_type>;
};
```
Or something like it?

cfjchu replied:
Thx for suggestion. I've added it and constraining `extract_locals` w/ new concept.

ayerofieiev-tt commented on tt_metal/api/tt-metalium/maybe_remote.hpp:
```diff
@@ -0,0 +1,237 @@
+    [[nodiscard]] const T* operator->() const;
+
+    // Optional-style access
+    [[nodiscard]] std::optional<T> optional_value() const;
```
double check, maybe not used

cfjchu replied:
Not used. I've removed.

cfjchu committed:
Add MaybeRemoteLike concept and update extract_locals to use it

- Add MaybeRemoteLike concept to check if a type behaves like MaybeRemote
- Update extract_locals function to use the concept as a constraint
- Include <concepts> header for C++20 concept support

cfjchu committed:
Remove unused MaybeRemote::optional_value method

cfjchu committed:
Apply feedback

Denys88 commented on tt_metal/api/tt-metalium/maybe_remote.hpp:
```diff
@@ -0,0 +1,242 @@
+}
+
+template <typename T>
+T* MaybeRemote<T>::operator->() {
```
```
template <typename T>
T* MaybeRemote<T>::operator->() {
    if (auto* p = std::get_if<T>(&value_); p) {
        return p;                     // local -– hand back the pointer
    }
    throw_remote_access_error();      // remote – raise
}
```

Looks a little bit better

Also please don't forget to add nodiscard. cc @omilyutin-tt 
I dont want to enforce it everwhere in the ttnn code but for generic types I thnk it is a must.

tt-asaigal commented on tt_metal/distributed/coordinate_translation.cpp:
```diff
@@ -31,14 +32,17 @@ const MeshContainer<PhysicalMeshCoordinate>& get_system_mesh_coordinate_translat
 
         const auto mesh_id = mesh_ids.front();
-        const auto mesh_shape = control_plane.get_physical_mesh_shape(mesh_id);
+        const auto local_mesh_shape = control_plane.get_physical_mesh_shape(mesh_id, tt::tt_fabric::MeshScope::LOCAL);
```
Thinking of multi mesh here. `get_physical_mesh_shape` will call `routing_table_generator_->mesh_graph->get_mesh_shape()` with a host id specified.

In multi mesh scenarios, this ID will be 0 (if its only multi mesh). Does the mesh ID get ignored here and do we always look at just host rank 0?

cfjchu replied:
In the context of using multi-process/multi-host, each process will having a binding to a `MeshId` which does not have to be `0` which is a non-issue. 

This is only a problem with single-process multi-mesh and part of larger changes that we talked about that needs to happen with `SystemMesh`. This is an existing limitation already if you look at what's already here. This change doesn't attempt to solve that problem.

cfjchu committed:
Address PR feedback

cfjchu commented on tt_metal/distributed/system_mesh.cpp:
```diff
@@ -194,13 +241,15 @@ uint32_t SystemMesh::get_physical_mesh_id(const MeshCoordinate& coord) const {
 
-const MeshShape& SystemMesh::get_shape() const { return pimpl_->get_shape(); }
+const MeshShape& SystemMesh::shape() const { return pimpl_->shape(); }
+
```
What `SystemMesh` represents is the `Mesh` from the MGD. This represents the physically connected mesh of devices, and hence should represent the global view that span multiple hosts.

cfjchu commented on tt_metal/distributed/distributed_coordinate_system.cpp:
```diff
@@ -0,0 +1,91 @@
+    auto& control_plane = tt::tt_metal::MetalContext::instance().get_control_plane();
+
+    const auto local_mesh_id = control_plane.get_local_mesh_id_bindings()[0];
+    auto local_offset = control_plane.get_local_mesh_offset();
```
I agree. Though, let's do this outside the scope of this PR.

cfjchu committed:
Address remaining feedback

cfjchu commented on tt_metal/distributed/mesh_device.cpp:
```diff
@@ -122,7 +127,7 @@ MeshDevice::ScopedDevices::ScopedDevices(
     ScopedDevices(
         config.physical_device_ids().empty()
-            ? SystemMesh::instance().get_mapped_physical_device_ids(config.mesh_shape(), config.offset())
+            ? extract_locals(SystemMesh::instance().get_mapped_physical_device_ids(config.mesh_shape(), config.offset()))
```
done

cfjchu committed:
Fix clang-tidy issue
